File: app/src/main/java/com/majeur/psclient/JoinRoomDialog.java
package com.majeur.psclient;

import android.graphics.Typeface;
import android.os.Bundle;
import android.text.Editable;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.RelativeSizeSpan;
import android.text.style.StyleSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.EditText;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.majeur.psclient.model.RoomInfo;
import com.majeur.psclient.util.SimpleTextWatcher;

import java.util.Arrays;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;

public class JoinRoomDialog extends DialogFragment {

    private static final String ARG_OFFICIAL_ROOMS = "official-rooms";
    private static final String ARG_CHAT_ROOMS = "chat-rooms";

    public static JoinRoomDialog newInstance(RoomInfo[] officialRooms, RoomInfo[] chatRooms) {
        JoinRoomDialog joinRoomDialog = new JoinRoomDialog();
        Bundle bundle = new Bundle();
        bundle.putSerializable(ARG_OFFICIAL_ROOMS, officialRooms);
        bundle.putSerializable(ARG_CHAT_ROOMS, chatRooms);
        joinRoomDialog.setArguments(bundle);
        return joinRoomDialog;
    }

    private RoomInfo[] mOfficialRooms;
    private RoomInfo[] mChatRooms;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mOfficialRooms = (RoomInfo[]) getArguments().getSerializable(ARG_OFFICIAL_ROOMS);
        Arrays.sort(mOfficialRooms, RoomInfo.COMPARATOR);

        mChatRooms = (RoomInfo[]) getArguments().getSerializable(ARG_CHAT_ROOMS);
        Arrays.sort(mChatRooms, RoomInfo.COMPARATOR);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.dialog_join_room, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ListView listView = (ListView) view;
        listView.setAdapter(mListAdapter);
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                RoomInfo roomInfo = (RoomInfo) mListAdapter.getItem(i);
                joinRoom(roomInfo.name);

            }
        });
        View footerView = getLayoutInflater().inflate(R.layout.list_footer_other_room, listView, false);
        listView.addFooterView(footerView);
        final View joinButton = footerView.findViewById(R.id.button_join_room);
        final EditText editText = footerView.findViewById(R.id.other_room_input);
        joinButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String input = editText.getText().toString();
                if (input.startsWith("battle-")) {
                    Toast.makeText(getContext(), "You cannot join a battle from here", Toast.LENGTH_LONG).show();
                    return;
                }
                joinRoom(input);
            }
        });
        editText.addTextChangedListener(new SimpleTextWatcher() {
            @Override
            public void afterTextChanged(Editable editable) {
                joinButton.setEnabled(editable.length() > 0);
            }
        });
    }

    private void joinRoom(String roomId) {
        roomId = roomId.toLowerCase().replaceAll("[^a-z0-9-]", "").trim();
        MainActivity activity = (MainActivity) getActivity();
        activity.getService().sendGlobalCommand("join", roomId);
        dismiss();
    }

    private final ListAdapter mListAdapter = new BaseAdapter() {

        private static final int VIEW_TYPE_HEADER = 0;
        private static final int VIEW_TYPE_REGULAR = 1;

        @Override
        public int getCount() {
            return 2 + mOfficialRooms.length + mChatRooms.length;
        }

        @Override
        public boolean isEnabled(int position) {
            return getItemViewType(position) == VIEW_TYPE_REGULAR;
        }

        @Override
        public boolean areAllItemsEnabled() {
            return false;
        }

        @Override
        public int getViewTypeCount() {
            return 2;
        }

        @Override
        public int getItemViewType(int position) {
            if (position == 0 || position == mOfficialRooms.length+1)
                return VIEW_TYPE_HEADER;
            return VIEW_TYPE_REGULAR;
        }

        class ViewHolder {
            TextView titleView;
            TextView descrView;
            View iconView;
        }

        @Override
        public View getView(int i, View view, ViewGroup viewGroup) {
            ViewHolder viewHolder;
            if (view == null) {
                view = getLayoutInflater().inflate(R.layout.list_item_room, viewGroup, false);
                viewHolder = new ViewHolder();
                viewHolder.titleView = view.findViewById(R.id.title_text_view);
                viewHolder.descrView = view.findViewById(R.id.descr_text_view);
                viewHolder.iconView = view.findViewById(R.id.imageView);
                view.setTag(viewHolder);
            } else {
                viewHolder = (ViewHolder) view.getTag();
            }

            Object item = getItem(i);
            if (item instanceof String) {
                SpannableString spannableString = new SpannableString((String) item);
                spannableString.setSpan(new RelativeSizeSpan(1.35f), 0, spannableString.length(),
                        Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
                viewHolder.titleView.setText(spannableString);
                viewHolder.descrView.setVisibility(View.GONE);
                viewHolder.iconView.setVisibility(View.GONE);
            } else {
                RoomInfo roomInfo = (RoomInfo) item;
                SpannableString spannableString = new SpannableString(roomInfo.name + " (" +
                        roomInfo.userCount + ")");
                spannableString.setSpan(new StyleSpan(Typeface.BOLD), 0,
                        roomInfo.name.length(), Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
                spannableString.setSpan(new StyleSpan(Typeface.ITALIC), roomInfo.name.length()+1,
                        spannableString.length(), Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
                spannableString.setSpan(new RelativeSizeSpan(0.8f), roomInfo.name.length()+1,
                        spannableString.length(), Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
                viewHolder.titleView.setText(spannableString);
                viewHolder.descrView.setText(roomInfo.description);
                viewHolder.descrView.setVisibility(View.VISIBLE);
                viewHolder.iconView.setVisibility(View.VISIBLE);
            }

            return view;
        }

        @Override
        public Object getItem(int i) {
            int officialRoomsCount = mOfficialRooms.length;
            if (i == 0)
                return "Official Rooms";
            if (i < officialRoomsCount + 1)
                return mOfficialRooms[i-1];
            if (i == officialRoomsCount + 1)
                return "Chat Rooms";
            return mChatRooms[i - officialRoomsCount - 2];
        }

        @Override
        public long getItemId(int i) {
            return 0;
        }
    };

}

File: app/src/main/java/com/majeur/psclient/ui/JoinRoomDialog.kt
package com.majeur.psclient.ui

import android.graphics.Typeface
import android.os.Bundle
import android.text.Editable
import android.text.SpannableString
import android.text.Spanned
import android.text.style.RelativeSizeSpan
import android.text.style.StyleSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.DialogFragment
import com.majeur.psclient.R
import com.majeur.psclient.databinding.DialogJoinRoomBinding
import com.majeur.psclient.databinding.ListFooterOtherRoomBinding
import com.majeur.psclient.model.RoomInfo
import com.majeur.psclient.util.SimpleTextWatcher
import java.util.*


class JoinRoomDialog : DialogFragment() {

    private lateinit var officialRooms: List<RoomInfo>
    private lateinit var chatRooms: List<RoomInfo>

    private var _binding: DialogJoinRoomBinding? = null
    private val binding get() = _binding!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        officialRooms = requireArguments().getSerializable(ARG_OFFICIAL_ROOMS) as List<RoomInfo>
        chatRooms = requireArguments().getSerializable(ARG_CHAT_ROOMS) as List<RoomInfo>
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = DialogJoinRoomBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.list.adapter = listAdapter
        binding.list.onItemClickListener = AdapterView.OnItemClickListener { _, _, index, _ ->
            val roomInfo = listAdapter.getItem(index) as RoomInfo
            joinRoom(roomInfo.name)
        }
        val footerBinding = ListFooterOtherRoomBinding.inflate(layoutInflater, binding.list, false)
        footerBinding.button.setOnClickListener(View.OnClickListener {
            val input = footerBinding.roomNameInput.text.toString()
            if (input.startsWith("battle-", ignoreCase = true)) {
                Toast.makeText(context, "You cannot join a battle from here", Toast.LENGTH_LONG).show()
                return@OnClickListener
            }
            joinRoom(input)
        })
        footerBinding.roomNameInput.addTextChangedListener(object : SimpleTextWatcher() {
            override fun afterTextChanged(editable: Editable) {
                footerBinding.button.isEnabled = editable.isNotEmpty()
            }
        })
        binding.list.addFooterView(footerBinding.root)
    }

    private fun joinRoom(roomId: String) {
        val regex = "[^a-z0-9-]".toRegex()
        roomId.toLowerCase(Locale.ROOT).replace(regex, "").let {
            (requireActivity() as MainActivity).service?.sendGlobalCommand("join", roomId)
            dismiss()
        }
    }

    private val listAdapter: ListAdapter = object : BaseAdapter() {

        private val VIEW_TYPE_HEADER = 0
        private val VIEW_TYPE_REGULAR = 1

        override fun getCount() =  2 + officialRooms.size + chatRooms.size

        override fun isEnabled(position: Int) = getItemViewType(position) == VIEW_TYPE_REGULAR

        override fun areAllItemsEnabled() = false

        override fun getViewTypeCount() = 2

        override fun getItemViewType(position: Int): Int {
            return if (position == 0 || position == officialRooms.size + 1) VIEW_TYPE_HEADER else VIEW_TYPE_REGULAR
        }

        inner class ViewHolder {
            var titleView: TextView? = null
            var descrView: TextView? = null
            var iconView: View? = null
        }

        override fun getView(i: Int, view: View?, viewGroup: ViewGroup): View {
            val convertView: View
            val viewHolder: ViewHolder
            if (view == null) {
                convertView = layoutInflater.inflate(R.layout.list_item_room, viewGroup, false)
                viewHolder = ViewHolder()
                viewHolder.titleView = convertView.findViewById(R.id.title_text_view)
                viewHolder.descrView = convertView.findViewById(R.id.descr_text_view)
                viewHolder.iconView = convertView.findViewById(R.id.imageView)
                convertView.tag = viewHolder
            } else {
                convertView = view
                viewHolder = view.tag as ViewHolder
            }
            val item = getItem(i)
            if (item is String) {
                val spannableString = SpannableString(item)
                spannableString.setSpan(RelativeSizeSpan(1.35f), 0, spannableString.length,
                        Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
                viewHolder.titleView!!.text = spannableString
                viewHolder.descrView!!.visibility = View.GONE
                viewHolder.iconView!!.visibility = View.GONE
            } else {
                val roomInfo = item as RoomInfo
                val spannableString = SpannableString(roomInfo.name + " (" +
                        roomInfo.userCount + ")")
                spannableString.setSpan(StyleSpan(Typeface.BOLD), 0,
                        roomInfo.name.length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
                spannableString.setSpan(StyleSpan(Typeface.ITALIC), roomInfo.name.length + 1,
                        spannableString.length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
                spannableString.setSpan(RelativeSizeSpan(0.8f), roomInfo.name.length + 1,
                        spannableString.length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
                viewHolder.titleView!!.text = spannableString
                viewHolder.descrView!!.text = roomInfo.description
                viewHolder.descrView!!.visibility = View.VISIBLE
                viewHolder.iconView!!.visibility = View.VISIBLE
            }
            return convertView
        }

        override fun getItem(i: Int): Any {
            val officialRoomsCount = officialRooms.size
            if (i == 0) return "Official Rooms"
            if (i < officialRoomsCount + 1) return officialRooms[i - 1]
            return if (i == officialRoomsCount + 1) "Chat Rooms" else chatRooms[i - officialRoomsCount - 2]
        }

        override fun getItemId(i: Int) = 0L
    }

    companion object {
        private const val ARG_OFFICIAL_ROOMS = "official-rooms"
        private const val ARG_CHAT_ROOMS = "chat-rooms"
        fun newInstance(officialRooms: Array<RoomInfo?>?, chatRooms: Array<RoomInfo?>?): JoinRoomDialog {
            val joinRoomDialog = JoinRoomDialog()
            val bundle = Bundle()
            bundle.putSerializable(ARG_OFFICIAL_ROOMS, officialRooms)
            bundle.putSerializable(ARG_CHAT_ROOMS, chatRooms)
            joinRoomDialog.arguments = bundle
            return joinRoomDialog
        }
    }
}
